axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderLineGeneratorServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+import com.axelor.apps.sale.exception.SaleExceptionMessage;
+import com.axelor.apps.sale.service.saleorderline.SaleOrderLineComputeService;
+import com.axelor.apps.sale.service.saleorderline.SaleOrderLineDomainService;
+import com.axelor.apps.sale.service.saleorderline.SaleOrderLineInitValueService;
+import com.axelor.apps.sale.service.saleorderline.SaleOrderLineProductService;
+import com.axelor.i18n.I18n;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+
+public class SaleOrderLineGeneratorServiceImpl implements SaleOrderLineGeneratorService {
+  protected SaleOrderLineInitValueService saleOrderLineInitValueService;
+  protected SaleOrderLineProductService saleOrderLineProductService;
+  protected SaleOrderLineRepository saleOrderLineRepository;
+  protected ProductRepository productRepository;
+  protected SaleOrderRepository saleOrderRepository;
+  protected SaleOrderLineComputeService saleOrderLineComputeService;
+  protected SaleOrderComputeService saleOrderComputeService;
+  protected SaleOrderLineDomainService saleOrderLineDomainService;
+  protected SaleOrderService saleOrderService;
+  protected SaleOrderOnLineChangeService saleOrderOnLineChangeService;
+  protected AppBaseService appBaseService;
+
+  @Inject
+  public SaleOrderLineGeneratorServiceImpl(
+      SaleOrderLineInitValueService saleOrderLineInitValueService,
+      SaleOrderLineProductService saleOrderLineProductService,
+      SaleOrderRepository saleOrderRepository,
+      SaleOrderLineRepository saleOrderLineRepository,
+      SaleOrderLineComputeService saleOrderLineComputeService,
+      SaleOrderComputeService saleOrderComputeService,
+      SaleOrderLineDomainService saleOrderLineDomainService,
+      ProductRepository productRepository,
+      SaleOrderService saleOrderService,
+      AppBaseService appBaseService,
+      SaleOrderOnLineChangeService saleOrderOnLineChangeService) {
+    this.saleOrderLineInitValueService = saleOrderLineInitValueService;
+    this.saleOrderLineProductService = saleOrderLineProductService;
+    this.saleOrderLineRepository = saleOrderLineRepository;
+    this.saleOrderRepository = saleOrderRepository;
+    this.saleOrderLineComputeService = saleOrderLineComputeService;
+    this.saleOrderComputeService = saleOrderComputeService;
+    this.saleOrderLineDomainService = saleOrderLineDomainService;
+    this.productRepository = productRepository;
+    this.saleOrderService = saleOrderService;
+    this.saleOrderOnLineChangeService = saleOrderOnLineChangeService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Transactional(rollbackOn = {Exception.class})
+  @Override
+  public SaleOrderLine createSaleOrderLine(SaleOrder saleOrder, Product product)
+      throws AxelorException {
+    if (saleOrder == null || product == null) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_NO_VALUE,
+          I18n.get(SaleExceptionMessage.EITHER_PRODUCT_OR_SALE_ORDER_ARE_NULL));
+    }
```

Comment:
You can move it to a separate method and add another error. We can currently add a line to the sale order which is not in draft, which should not be possible. Please check if the sale order is in draft, otherwise thow an error : "You can only add a line to a draft sale order."